components/permissions/contexts/brave_ethereum_permission_context.h

```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_PERMISSIONS_CONTEXTS_BRAVE_ETHEREUM_PERMISSION_CONTEXT_H_
+#define BRAVE_COMPONENTS_PERMISSIONS_CONTEXTS_BRAVE_ETHEREUM_PERMISSION_CONTEXT_H_
+
+#include <map>
+#include <queue>
+#include <string>
+
+#include "components/permissions/permission_context_base.h"
+#include "components/permissions/permission_request_id.h"
+
+class GURL;
+
+namespace content {
+class BrowserContext;
+class WebContents;
+}  // namespace content
+
+namespace permissions {
+
+class BraveEthereumPermissionContext : public PermissionContextBase {
+ public:
+  // using PermissionContextBase::RequestPermission;
+  explicit BraveEthereumPermissionContext(
+      content::BrowserContext* browser_context);
+  ~BraveEthereumPermissionContext() override;
+
+  BraveEthereumPermissionContext(const BraveEthereumPermissionContext&) =
+      delete;
+  BraveEthereumPermissionContext& operator=(
+      const BraveEthereumPermissionContext&) = delete;
+
+  /**
+   * This is called by PermissionManager::RequestPermissions, for each
+   * permission request ID, we will parse the requesting_frame URL to get the
+   * ethereum address list to be used for each sub-request. Each sub-request
+   * will then consume one address from the saved list and call
+   * PermissionContextBase::RequestPermission with it.
+   */
+  void RequestPermission(content::WebContents* web_contents,
+                         const PermissionRequestID& id,
+                         const GURL& requesting_frame,
+                         bool user_gesture,
+                         BrowserPermissionCallback callback) override;
+
```

Comment:
I haven't tried it myself, but why not override ```std::unique_ptr<permissions::PermissionRequest> CreatePermissionRequest``` and do all ethereum-specific permission stuff in ```BraveEthereumPermissionContext```? Maybe it's possible to not touch ```PermissionContextBase``` at all?
see ```FileHandlingPermissionContext```.

Reply:
@goodov I thought we'll need to patch permission_context.h to add `BraveEthereumPermissionContext` as a friend class so we can access the private member we're using, is it possible to add it as a friend without patching?

Reply:
oh nvm, I forgot there was already a patch for permission_context_base.h, I'll try this out.

Reply:
Ah, now i remember why i put it in the PermissionContextBase when i implemented it, cuz they bind PermissionContextBase::PermissionDecided and PermissionContextBase::Cleanup directly when calling `CreatePermissionRequest`. 
Not sure what you meant by override `CreatePermissionRequest`, what's that for?

Reply:
> Not sure what you meant by override ```CreatePermissionRequest```, what's that for?

from what I see it's possible to create a custom request handler and somehow replace/append any logic to ```permission_decided_callback``` (you can store this callback, bind a new one and then call the stored one from your code). Not sure about ```Cleanup``` though (maybe it can be worked out somehow?)

it's also possible to override this in a custom handler:
```
  void PermissionGranted(bool is_one_time) override;
  void PermissionDenied() override;
```
```DownloadPermissionRequest``` does something custom. It maybe possible to inherit a handler from ```PermissionRequestImpl``` so you can reuse everything and then just add our specific bits.

Reply:
I can't find a way to achieve the same without touching `PermissionContextBase`, as discussed in DM, I'll keep it as is.